### PR TITLE
sessions decoupling from discourse

### PR DESCRIPTION
**Pull request by orlando**
This PR refactors authentication. The goal is to reduce API calls to Discourse and decouple the app a bit from it.

I tried to keep this PR just to remove the sessions, but I changed some parts of the app to remove unnecessary calls to Discourse.

Here I list the changes https://gist.github.com/orlando/64b924034b3a649a80bd44badd811777, I didn't include specific implementation since it was more of adapting what we have to use Passport. The passport integration with Discourse SSO was done in `lib/passport` folder.

Most of the problems I had was with testing, I removed some logging in order to make failures more readable. Also, we are hiding too much of the implementation of these tests, which make them hard to debug.

Fixes https://github.com/debtcollective/parent/issues/190, https://github.com/debtcollective/parent/issues/167, https://github.com/debtcollective/parent/issues/191

### CODE REVIEW

**Comment by orlando** [comment]
The tests are failing due Circle CI not copying the config file. Looks like the code is there (https://github.com/debtcollective/dispute-tools/blob/development/.circleci/config.yml#L24) but is not working for some reason

File: controllers/SessionsController.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+const Raven = require('raven');
+const passport = require('$lib/passport');
+const sso = require('$services/sso');
+const config = require('$config/config');
+
+const SessionsController = Class('SessionsController').inherits(BaseController)({
+  prototype: {
+    create(req, res) {
+      if (req.user) {
+        req.flash('info', 'You are already logged in');
```
Comment (marcondag) [inline controllers/SessionsController.js]: I don't think we need to add this message. The user doesn't really care either way, they just want to be logged in so we can just redirect them.

File: controllers/SessionsController.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/* globals Class, BaseController, User, UserMailer */
```
Comment (marcondag) [inline controllers/SessionsController.js]: Can you fix these globals? Everything except `Class` can be required in from the actual file (and `UserMailer` doesn't even exist anymore so we can just remove it).

File: controllers/SessionsController.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+const Raven = require('raven');
```
Comment (marcondag) [inline controllers/SessionsController.js]: We should import Raven from `$lib`

File: controllers/SessionsController.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          }
+
+          req.flash('success', 'Welcome to The Debt Collective');
+
+          return req.session.save(() => {
+            Raven.setContext({
+              user: {
+                email: user.email,
+                id: user.id,
+                external_id: user.external_id,
```
Comment (marcondag) [inline controllers/SessionsController.js]: Can we use camel case for the property name?

File: lib/passport/discourse.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+const PassportStrategy = require('passport-strategy');
+const sso = require('$services/sso');
+const Raven = require('raven');
```
Comment (marcondag) [inline lib/passport/discourse.js]: Also import it from `$lib` here

File: lib/passport/discourse.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      try {
+        user = await sso.handlePayload(payload);
+      } catch (e) {
+        Raven.captureException(e);
+        return this.fail('handlePayload error');
+      }
+
+      this.success(user);
+    } else {
+      this.fail('Invalid sig');
```
Comment (marcondag) [inline lib/passport/discourse.js]: Should this error be something more like `Invalid parameters` or something? If we want a more specific error from the sso we can change `extractPayload` and `handlePayload` to throw more specific instances of `AuthenticationError` and then we can log invalid signatures?

File: lib/passport/index.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+/* globals User, CONFIG */
```
Comment (marcondag) [inline lib/passport/index.js]: Let's also import these variables from their files.

File: middlewares/SafeLocals.js [hunk 8]
```diff
@@ -1,8 +1,7 @@
 /* globals CONFIG, Collective */
```
Comment (marcondag) [inline middlewares/SafeLocals.js]: We can remove these

File: middlewares/flashMessages.js [hunk 9]
```diff
@@ -1,28 +1 @@
...
-        flash: JSON.stringify(res.locals.flash),
-      })}`;
-      redirect.call(this, location, ...args);
-    } else {
-      redirect.call(this, location, ...args);
-    }
-  };
-  next();
-};
+module.exports = require('req-flash')({ locals: 'flash' });
```
Comment (marcondag) [inline middlewares/flashMessages.js]: God, I'm glad we can get rid of that hack

File: migrations/20180912161052_add_profile_fields_to_user.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+exports.up = knex =>
+  knex.schema.table('Users', t => {
+    t.string('name').defaultTo('');
+    t.string('email');
+    t.string('username');
+    t.boolean('admin').defaultTo(false);
+    t.string('avatar_url');
+  });
```
Comment (marcondag) [inline migrations/20180912161052_add_profile_fields_to_user.js]: Would we potentially want to store the user's groups for any reason? Might be useful for querying from metabase (unless metabase grabs up-to-date user info from the discourse db)?

File: models/User.js [hunk 11]
```diff
@@ -1,23 +1,62 @@
...
       Krypton.Model.prototype.init.call(this, config);
     },
     setInfo(info) {
-      return Object.assign(this, {
-        ...info,
-        // Discourse will return an object with an id,
-        // so this way we make sure our database id doesn't get overwritten
-        id: this.id,
-        external_id: info.id,
+      const hasAdminRole = _.includes(info.groups, config.discourse.adminRole);
```
Comment (marcondag) [inline models/User.js]: Can we just use native Array.includes?

File: package.json [hunk 12]
```diff
@@ -26,13 +30,12 @@
...
     "utils:knexreset": "scripty",
     "utils:copybadges": "scripty",
-    "migrate:dev": "knex migrate:latest --env development",
-    "migrate:seed:dev": "knex seed:run --env development",
     "db:migrate": "knex migrate:latest",
+    "db:rollback": "knex migrate:rollback",
     "db:seed": "knex seed:run",
     "report": "yarn istanbul-combine -d coverage -p summary -r lcov coverage/{unit,integration,form-validations}/coverage.raw.json",
-    "start": "scripty",
-    "start:dev": "nodemon --inspect=9220 bin/server.js -i src/**/*",
```
Comment (marcondag) [inline package.json]: Can you leave the `start:dev` script? I use it quite a bit.

File: services/sso.js [hunk 13]
```diff
@@ -1,13 +1,13 @@
-/* globals CONFIG, User */
+/* globals User */
```
Comment (marcondag) [inline services/sso.js]: If we can fix this import too.

File: tests/integration/controllers/Admin.DisputesControllerTest.js [hunk 14]
```diff
@@ -242,14 +218,6 @@ describe('Admin.DisputesController', () => {
       describe('when admin', () => {
         it('should allow', () => testAllowed(testGet(url, admin)));
       });
-
-      describe('when dispute admin', () => {
```
Comment (marcondag) [inline tests/integration/controllers/Admin.DisputesControllerTest.js]: Can we leave the authentication tests?

File: tests/utils/index.js [hunk 15]
```diff
@@ -39,51 +51,20 @@ const helpers = {
...
-    params = { externalId: ids.nextExternal() },
-    groups = [],
-    admin = false,
-    moderator = false,
-  } = {}) {
-    const user = new User(params);
-
-    await user.save();
-
-    nock(baseUrl)
```
Comment (marcondag) [inline tests/utils/index.js]: Removing this is why that test is failing.

File: controllers/SessionsController.js (hunk 1 above)
Comment (orlando) [reply controllers/SessionsController.js]: 👍

File: migrations/20180912161052_add_profile_fields_to_user.js (hunk 10 above)
Comment (orlando) [reply migrations/20180912161052_add_profile_fields_to_user.js]: Yeah, we will have both databases in metabase, but joins between databases is not supported. We can work around this by using [foreign data wrappers](https://robots.thoughtbot.com/postgres-foreign-data-wrapper) in postgres if the necessity comes

**Comment by orlando** [comment]
@marcondag changes up and tests green

**Pull request merged**